app/scripts/migrations/140.ts

```diff
@@ -0,0 +1,251 @@
+ * @param state - The MetaMask extension state.
+ * @returns The updated MetaMask extension state.
+ */
+function transformState(state: Record<string, unknown>) {
+  const {
+    NetworkController: networkControllerState,
+    PermissionController: permissionControllerState,
+    SelectedNetworkController: selectedNetworkControllerState,
+  } = state;
+  if (!networkControllerState || !isObject(networkControllerState)) {
+    log.warn(
+      'Skipping migration: `NetworkController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (
+    !hasProperty(networkControllerState, 'selectedNetworkClientId') ||
+    typeof networkControllerState.selectedNetworkClientId !== 'string'
+  ) {
+    log.warn(
+      'Skipping migration: `NetworkController.selectedNetworkClientId` is not a string.',
+    );
+    return state;
+  }
+
+  if (
+    !hasProperty(networkControllerState, 'networkConfigurationsByChainId') ||
+    !isObject(networkControllerState.networkConfigurationsByChainId)
+  ) {
+    log.warn(
+      'Skipping migration: `NetworkController.networkConfigurationsByChainId` is not an object.',
+    );
+    return state;
+  }
+
+  if (!permissionControllerState || !isObject(permissionControllerState)) {
+    log.warn(
+      'Skipping migration: `PermissionController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(permissionControllerState.subjects)) {
+    log.warn(
+      'Skipping migration: `PermissionController.subjects` state is not an object.',
+    );
+    return state;
+  }
+
+  if (
+    !selectedNetworkControllerState ||
+    !isObject(selectedNetworkControllerState)
+  ) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController` state not found or is not an object.',
+    );
+    return state;
+  }
+
+  if (!isObject(selectedNetworkControllerState.domains)) {
+    log.warn(
+      'Skipping migration: `SelectedNetworkController.domains` state is not an object.',
+    );
+    return state;
+  }
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+  if (typeof currentChainId !== 'string') {
+    log.warn(
+      'Skipping migration: `NetworkController.selectedNetworkClientId` is not a string.',
+    );
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration =
+      permissionKeys.includes(SnapEndowments.EthereumProvider) &&
```

Comment:
I'm assuming this permission can only be granted to Snaps? Otherwise we could be overwriting permissions for non-snaps subjects?

Reply:
Yes, it can only be granted to Snaps.